tests/test_appipe.py

```diff
@@ -21,22 +21,142 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
 
+import contextlib
+import os
 import unittest
+from unittest.mock import patch, Mock
+
 import lsst.utils.tests
+import lsst.pex.exceptions as pexExcept
+import lsst.daf.persistence as dafPersist
+import lsst.pipe.base as pipeBase
+
+from lsst.ap.pipe import ApPipeTask
 
 
 class PipelineTestSuite(lsst.utils.tests.TestCase):
     '''
     A set of tests for the functions in ap_pipe.
-
-    TODO: write tests for DM-13567.
     '''
 
-#    def testDoProcessCcd(self):
-#        # test something
+    @classmethod
+    def _makeDefaultConfig(cls):
+        config = ApPipeTask.ConfigClass()
+        config.load(os.path.join(cls.datadir, "config", "apPipe.py"))
+        config.ppdb.db_url = "sqlite://"
+        config.ppdb.isolation_level = "READ_UNCOMMITTED"
+        return config
+
+    @classmethod
+    def setUpClass(cls):
+        try:
+            cls.datadir = lsst.utils.getPackageDir("ap_pipe_testdata")
+        except pexExcept.NotFoundError:
+            raise unittest.SkipTest("ap_pipe_testdata not set up")
+
+        cls.config = cls._makeDefaultConfig()
+
+    def setUp(self):
+        self.butler = dafPersist.Butler(inputs={'root': self.datadir})
+
+        def makeMockDataRef(datasetType, level=None, dataId={}, **rest):
+            mockDataRef = Mock(dafPersist.ButlerDataRef)
+            mockDataRef.dataId = dict(dataId, **rest)
+            return mockDataRef
+        butlerPatcher = patch.object(self.butler, "dataRef",
+                                     side_effect=makeMockDataRef)
+        butlerPatcher.start()
+        self.addCleanup(butlerPatcher.stop)
+        self.dataId = {"visit": 413635, "ccdnum": 42}
+
+    @contextlib.contextmanager
+    def mockPatchSubtasks(self, task):
+        """Make mocks for all the ap_pipe subtasks.
+
+        Parameters
+        ----------
+        task : `lsst.ap.pipe.ApPipeTask`
+            the task whose subtasks will be mocked
+
+        Yields
+        ------
+        subtasks : `lsst.pipe.base.Struct`
+            All mocks created by this context manager, including:
+
+            ``ccdProcessor``
+            ``differencer``
+            ``dpddifier``
+            ``associator``
+            ``forcedSource``
+                a mock for the corresponding subtask. Mocks do not return any
+                particular value, but have mocked methods that can be queried
+                for calls by ApPipeTask
+        """
+        with patch.object(task, "ccdProcessor") as mockCcdProcessor:
+            with patch.object(task, "differencer") as mockDifferencer:
+                with patch.object(task, "diaSourceDpddifier") as mockDpddifier:
+                    with patch.object(task, "associator") as mockAssociator:
+                        with patch.object(task, "diaForcedSource") as mockForcedSource:
```

Comment:
Ouch. You don't need to nest these, you can put them all in one context: `with patch1() as mock1, patch2() as mock2, ...: doStuff()`.

Reply:
Thank you, that works!